sensor/common/registry/lazy_tls_factory.go

```diff
@@ -0,0 +1,95 @@
+package registry
+
+import (
+	pkgErrors "github.com/pkg/errors"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/registries"
+	"github.com/stackrox/rox/pkg/registries/docker"
+	dockerFactory "github.com/stackrox/rox/pkg/registries/docker"
+	rhelFactory "github.com/stackrox/rox/pkg/registries/rhel"
+	"github.com/stackrox/rox/pkg/registries/types"
+)
+
+var (
+	lazyEligibleCreators = []types.CreatorWrapper{
+		dockerFactory.CreatorWithoutRepoList,
+		rhelFactory.CreatorWithoutRepoList,
+	}
+)
+
+var _ registries.Factory = (*lazyFactory)(nil)
+
+type lazyFactory struct {
+	creators      map[string]types.Creator
+	tlsCheckCache *tlsCheckCacheImpl
+}
+
+func newLazyFactory(tlsCheckCache *tlsCheckCacheImpl) registries.Factory {
+	factory := &lazyFactory{
+		creators:      make(map[string]types.Creator, len(lazyEligibleCreators)),
+		tlsCheckCache: tlsCheckCache,
+	}
+
+	for _, creatorFunc := range lazyEligibleCreators {
+		typ, creator := creatorFunc()
+		factory.creators[typ] = creator
+	}
+
+	return factory
+}
+
+// CreateRegistry performs aggressive up front validation so that errors can
+// be surfaced, otherwise errors may be lost during lazy initialization.
+func (e *lazyFactory) CreateRegistry(source *storage.ImageIntegration, options ...types.CreatorOption) (types.ImageRegistry, error) {
+	if source == nil {
+		return nil, pkgErrors.New("image integration is nil")
+	}
+
+	creator, exists := e.creators[source.GetType()]
+	if !exists {
+		return nil, pkgErrors.Errorf("registry with type '%s' does not exist", source.GetType())
+	}
+
+	dockerConfig, err := extractDockerConfig(source)
+	if err != nil {
+		// Only integrations with a docker config are eligible for lazy tls checking
+		return nil, pkgErrors.Wrap(err, "extracting docker config")
+	}
+
+	hostname, url := docker.RegistryHostnameURL(dockerConfig.GetEndpoint())
```

Comment:
For the aggressive up front validation, I recommend we add checks for spaces, be it leading/trailing or in the text.
We've recently had an issue where the registry URL contained spaces which lead to unwanted behavior.
WDYT?

Reply:
David knows about the case :) this PR is a reaction to the findings from it.

Reply:
The changes made in https://github.com/stackrox/stackrox/pull/11668 should address the spaces in hostname issue (for detected pull secrets): Spaces at the edges of the URL are [removed before 'upserting'](https://github.com/stackrox/stackrox/blob/master/sensor/kubernetes/listener/resources/secrets.go#L264) into the registry store and the TLS check was enhanced to [catch other url issues](rox/blob/master/pkg/tlscheck/tlscheck.go#L24).  So we should be good.

Happy to add another validation if is a strong opinion - but it would be unnecessary current state.

Reply:
Thanks for reminding me of 11668!
Between that PR and the upserting trim, I think we're good.